docs/testing/index.md

```diff
@@ -118,25 +117,256 @@ Link: CACAMBER
+* Mock
+* Stub
+* Faker
+* Spy
+
+#### Mock
+
+Ein Mock-Objekt ist ein Objekt, dass dynamisch auf die Eingaben der Aufrufers reagieren kann. Ein Mock-Objekt kann zum Beispiel eine Abfrage in einem externen System oder der Datenbank imitieren und testfallabhängig die gewünschten Daten liefern. 
+
+#### Stub
+
+Ein Stub ist ein Objekt mit minimaler Implementierung der zum Testen notwendigen Methoden. Dieses Objekt ist nur dazu da, um den fehlerfreien Aufruf der zu testenden Methoden zu gewährleisten. 
+
+#### Fake
+
+Ein Fake-Objekt liefert notwendige Daten nach einem vereinfachten Algorithmus. Beispielsweise könnte das Echte Objekt eine umfangreiche Postleitzahlenprüfung vornehmen, bei der Geodaten, Ortsteile und Postleitzahlen auf Grundlage verschiedener Dienste verifiziert werden. Der entsprechende Faker könnte einfache Prüfungen machen, die für die Durchführung der Tests ausreichend sind.
+
+#### Spy
+
+Ein Spion-Objekt kann Eigenschaften von Stubs, Fakes und Mocks enthalten, übernimmt jedoch zusätzlich noch die Funktion, Zugriffe zu protokollieren. So könnte ein Spy zum Beispiel beim Testen aufgerufen werden; es wird jedoch nicht direkt das Ergebnis geprüft, sondern nur, ob diese Methode aufgerufen wurde. Dieser Typ wird gerne verwendet, um sicherzustellen, dass bei bestimmten Aktionen E-Mails oder andere Nachrichten verschickt worden wären. 
+
+#### Links
+
+* [Martin Fowler - Mocks Aren't Stubs](https://www.martinfowler.com/articles/mocksArentStubs.html)
+
+### Test-Doubles
+
+Die Verwendung von Test-Doubles ist notwendig, wenn Abhängigkeiten bestehen, die nicht ausreichend aufgelöst wurden oder aufgelöst werden konnten. Mit entsprechenden Test-Double-Frameworks kann das Ergebnis von Datenbankzugriffe oder Funktionsbausteinaufrufen gefälscht werden. 
+
+Test-Double-Frameworks sind in der Regel umständlich zu bedienen und sehr unübersichtlich. Mit vielen Definitionen und Methoden müssen Eingabeparameter und die gewünschten Ergebnisse vorgegeben werden. Wenn möglich sollten Sie die Abhängigkeiten eliminieren um auf die Verwendung von Test-Doubles verzichten zu können. Dies ist jedoch nicht immer möglich. Deswegen stehen Test-Double-Frameworks für folgende Objekte zur Verfügung:
+* Datenbankzugriffe (oSQL)
+* Funktionsbausteine
+
+#### Test-Double-Framework für Datenbankzugriffe
+
+tbd
+
+#### Test-Double-Framework für Funktionsbausteine
+
+tbd
+
+### Test-Seams
+
+TODO: Variante der Test-Doubles oder eigene Technik?
+
+
 ### Testumgebung
-* Ausführen Unit Tests
-* Anzeige Ergebnis
-* Codeabdeckung
-* 
+
+Die ABAP-Unit-Tests können aus der SAP-Entwicklungsumgebung (SE80, SE24) oder den ABAP-Development-Tools heraus verwendet werden. Die Vorgehensweise unterscheidet sich nur in Kleinigkeiten. Die Techniken, die zur Erstellung notwendig sind, ähneln sich jedoch stark. Unit Test, die in der SE80 erstellt wurden, können auch in Eclipse gewartet und getestet werden und umgekehrt. Die Tastenkombinationen sind in beiden Tools gleich.

```

Comment:
Die Tastenkombinationen sind meines Wissens nicht ganz identisch, außer vielleicht Ctrl+Shift+F10 zur Ausführung. Für mit Coverage gibt's auch z. B. gar keine in der SE80, oder?

### ADT

- Ctrl + Shift +
  - F9: Unit Test Preview anzeigen
  - F10: Unit Tests ausführen
  - F11: Unit Tests mit Coverage ausführen
  - F12: Unit Test Ausführungsdialog aufrufen
  - (F2: ATC-Prüfung mit Standardvariante ausführen)

### ABAP Workbench

- Ctrl + Shift +
  - F10: Unit Tests ausführen
  - F11: Lokale Testklassen anzeigen (nur formularbasierter Editor)
- Ctrl +
  - F11: Lokale Testklassen anzeigen (nur Quelltext-basierter Editor)